internal/controllers/gateway/gateway_controller.go

```diff
@@ -582,6 +619,10 @@ func (r *GatewayReconciler) updateAddressesAndListenersStatus(
 		setGatewayCondition(gateway, readyCondition)
 		return true, r.Status().Update(ctx, pruneGatewayStatusConds(gateway))
```

Comment:
Maybe we should not return here. When we return here, the gateway status needs two rounds of update when the condition `ready` in gateway status was not updated.

Reply:
Maybe I'm missing something here? My understanding is we'll only see this update twice if the Listerner statuses change between updates, since we are setting `gateway.Status.Listeners = listenerStatuses` in the first return block. The next reconcile should then generate the same set, see they're equal, and do nothing.